homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,593 @@
+                zone_distance = distance(
+                    device.location["latitude"],
+                    device.location["longitude"],
+                    zone_state_lat,
+                    zone_state_long,
+                )
+                distances.append(round(zone_distance / 1000, 1))
+
+            if distances:
+                mindistance = min(distances)
+            else:
+                continue
+
+            # Calculate out how long it would take for the device to drive
+            # to the nearest zone at 120 km/h:
+            interval = round(mindistance / 2, 0)
+
+            # Never poll more than once per minute
+            interval = max(interval, 1)
+
+            if interval > 180:
+                # Three hour drive?
+                # This is far enough that they might be flying
+                interval = self._max_interval
+
+            if (
+                device.battery_level is not None
+                and device.battery_level <= 33
+                and mindistance > 3
+            ):
+                # Low battery - let's check half as often
+                interval = interval * 2
+
+            intervals[device.name] = interval
+
+        return max(
+            int(min(intervals.items(), key=operator.itemgetter(1))[1]),
+            self._max_interval,
+        )
+
+    def keep_alive(self, now):
+        """Keep the API alive."""
+        if self.api is None:
+            self.reset_account()
+
+        if self.api is None:
+            return
+
+        self.api.authenticate()
+        self.update_devices()
+
+    @property
+    def name(self):
+        """Return the account name."""
+        return self._accountname
+
+
+class IcloudDevice(Entity):
+    """Representation of a iCloud device."""
+
+    def __init__(self, account: IcloudAccount, device: AppleDevice):
+        """Initialize the iCloud device."""
+        self.hass = account.hass
+        self._account = account
+        self._accountname = account.name
+        self._accountname_slug = slugify(self._accountname.partition("@")[0])
+
+        self._device = device
+        self._status = device.status(DEVICE_STATUS_SET)
+        _LOGGER.debug("Device Status is %s", self._status)
+
+        self._name = self._status["name"]
+        self._dev_id = slugify(self._name.replace(" ", "", 99))  # devicename
+        self._unique_id = f"{self._accountname_slug}_{self._dev_id}"
+        self._device_class = self._status["deviceClass"]
+        self._device_name = self._status["deviceDisplayName"]
+        if self._status["prsId"]:
+            self._owner_fullname = account.family_members_fullname[
+                self._status["prsId"]
+            ]
+        else:
+            self._owner_fullname = account.account_owner_fullname
+
+        self._battery_level = None
+        self._battery_status = None
+        self._low_power_mode = None
+        self._location = None
+
+        self._seen = False
+
+        self.update(self._status)
```

Comment:
Please call this outside after creating the instance. Remove it from init method.